bundles/org.openhab.binding.monopriceaudio/src/main/java/org/openhab/binding/monopriceaudio/internal/handler/MonopriceAudioHandler.java

```diff
@@ -0,0 +1,723 @@
+            sourcesLabels.add(new StateOption("5", config.inputLabel5));
+            sourcesLabels.add(new StateOption("6", config.inputLabel6));
+
+            // Put the source labels on all active zones
+            List<Integer> activeZones = IntStream.range(1, numZones + 1).boxed().collect(Collectors.toList());
+
+            stateDescriptionProvider.setStateOptions(
+                    new ChannelUID(getThing().getUID(), ALL + CHANNEL_DELIMIT + CHANNEL_TYPE_ALLSOURCE), sourcesLabels);
+            activeZones.forEach(zoneNum -> {
+                stateDescriptionProvider.setStateOptions(new ChannelUID(getThing().getUID(),
+                        ZONE.toLowerCase() + zoneNum + CHANNEL_DELIMIT + CHANNEL_TYPE_SOURCE), sourcesLabels);
+            });
+
+            // remove the channels for the zones we are not using
+            if (numZones < MAX_ZONES) {
+                List<Channel> channels = new ArrayList<>(this.getThing().getChannels());
+
+                List<Integer> zonesToRemove = IntStream.range(numZones + 1, MAX_ZONES + 1).boxed()
+                        .collect(Collectors.toList());
+
+                zonesToRemove.forEach(zone -> {
+                    channels.removeIf(c -> (c.getUID().getId().contains(ZONE.toLowerCase() + zone)));
+                });
+                updateThing(editThing().withChannels(channels).build());
+            }
+
+            // initialize the all volume state
+            allVolume = initialAllVolume;
+            long allVolumePct = Math
+                    .round((double) (initialAllVolume - MIN_VOLUME) / (double) (MAX_VOLUME - MIN_VOLUME) * 100.0);
+            updateState(ALL + CHANNEL_DELIMIT + CHANNEL_TYPE_ALLVOLUME,
+                    new PercentType(BigDecimal.valueOf(allVolumePct)));
+
+            updateStatus(ThingStatus.UNKNOWN);
+
+            scheduleReconnectJob();
+            schedulePollingJob();
+        }
+    }
+
+    @Override
+    public void dispose() {
+        cancelReconnectJob();
+        cancelPollingJob();
+        closeConnection();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        String channel = channelUID.getId();
+        String[] channelSplit = channel.split(CHANNEL_DELIMIT);
+        MonopriceAudioZone zone = MonopriceAudioZone.valueOf(channelSplit[0].toUpperCase());
+        String channelType = channelSplit[1];
+
+        if (getThing().getStatus() != ThingStatus.ONLINE) {
+            logger.debug("Thing is not ONLINE; command {} from channel {} is ignored", command, channel);
+            return;
+        }
+
+        boolean success = true;
+        synchronized (sequenceLock) {
+            if (!connector.isConnected()) {
+                logger.debug("Command {} from channel {} is ignored: connection not established", command, channel);
+                return;
+            }
+
+            Stream<String> zoneStream = MonopriceAudioZone.VALID_ZONES.stream();
```

Comment:
Might as well save later code from having to call it.
```suggestion
            Stream<String> zoneStream = MonopriceAudioZone.VALID_ZONES.stream().limit(numZones);
```